### PR TITLE
Fix crackle in local audio

### PR DESCRIPTION
**Pull request by zzmp**
Sizes local injector buffer to network period, instead of device period, so that device callbacks are less likely to cause an underrun.

### CODE REVIEW

**Comment by zzmp** [comment]
### QA Plan

Play a sound with a local injector, and do something else (move, build, whatever).
In `master`, the sound will crackle occasionally. (It is a threading issue, so when it does so is unpredictable.)
In this PR, it should no longer crackle.

For reference, this is how to load your own local injector:
```
// this dither is a good test sound, as you will easily notice any crackle
var dither = SoundCache.getSound("http://xiph.org/~xiphmont/demo/1kHz-0-dither.wav");
// wait a minute for the dither to load, then...
Audio.playSound(dither, { position: MyAvatar.position, volume: 0.1 });
```

**Comment by zzmp** [comment]
(This includes work from #9936.)

**Comment by Nex-Pro** [comment]
Testing this @zzmp and @kencooke

**Comment by hifi-gustavo** [comment]
Build console output for commit: b6f652e0eae8289dfb75f2af39da3f9172eb8b56 available at: https://gist.github.com/2056d23b86e08c79b6b91d0d4ad21a7d<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10414-b6f652e.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10414-b6f652e.dmg

**Comment by Nex-Pro** [comment]
testing

**Comment by Nex-Pro** [comment]
### FAIL

With the same sound file it crackles almost as bad as before even when stationary and nothing else using my sound system.
When I loop the file in vlc media player there is zero crackle.

**Comment by Nex-Pro** [comment]
Moving makes it worse.

**Comment by zzmp** [comment]
@kencooke can you think of a source for crackling audio besides the local injectors' buffer not having the next sample ready to go?

**Comment by kencooke** [comment]
Feeding the callback with less than requested number of samples
Taking too long in the callback (mallocs, locks, etc)
Any mishandling of the audio buffers (dropping or adding samples, etc)

File: libraries/audio-client/src/AudioClient.cpp [hunk 1]
```diff
@@ -1547,6 +1570,9 @@ bool AudioClient::switchOutputToAudioDevice(const QAudioDeviceInfo& outputDevice
                     qCDebug(audioclient) << "local buffer (samples):" << localPeriod;
 
                     disconnect(_audioOutput, &QAudioOutput::stateChanged, 0, 0);
+
+                    // unlock to avoid a deadlock with the device callback (which always succeeds this initialization)
```
Comment (zzmp) [inline libraries/audio-client/src/AudioClient.cpp]: @kencooke I do not know if this assumption is correct,
> which always succeeds this initialization

It was fine in my testing, and I want to assume this so I can avoid using a recursive mutex. Do you know about this?

File: libraries/audio-client/src/AudioClient.cpp [hunk 2]
```diff
@@ -1685,12 +1711,24 @@ qint64 AudioClient::AudioOutputIODevice::readData(char * data, qint64 maxSize) {
...
-        samplesRequested = std::min(samplesRequested, _audio->_localSamplesAvailable.load(std::memory_order_acquire));
+        // - switchOutputToAudioDevice will zero samples count,
+        //   stop the device - so that readData will exhaust the existing buffer or see a zeroed samples count,
+        //   and start the device - which can then only see a zeroed samples count
+        int samplesAvailable = _audio->_localSamplesAvailable.load(std::memory_order_acquire);
+
+        // if we do not have enough samples buffered despite having injectors, buffer them synchronously
+        if (samplesAvailable < samplesRequested && _audio->_localInjectorsAvailable.load(std::memory_order_acquire)) {
+            // try_to_lock, in case the device is being shut down already
+            std::unique_ptr<Lock> localAudioLock(new Lock(_audio->_localAudioMutex, std::try_to_lock));
```
Comment (zzmp) [inline libraries/audio-client/src/AudioClient.cpp]: This is the crux of this fix.

**Comment by hifi-gustavo** [comment]
Build console output for commit: c478f1a7520d7ef0d4901eaaa2726bd6b043a24d available at: https://gist.github.com/2056d23b86e08c79b6b91d0d4ad21a7d<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10414-c478f1a.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10414-c478f1a.dmg

**Comment by zzmp** [comment]
The issue was that the injectors buffer did not always have enough samples to feed the callback. I confirmed this with debug statements locally.

To fix it, I am synchronously filling the buffer from within the device callback in the cases where it would otherwise starve.

**Comment by Nex-Pro** [comment]
Awaiting build

**Comment by hifi-gustavo** [comment]
Build console output for commit: 219062b76a028651ee437fcd64ae9c7f0172c754 available at: https://gist.github.com/2056d23b86e08c79b6b91d0d4ad21a7d<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10414-219062b.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10414-219062b.dmg

**Comment by Nex-Pro** [comment]
Testing this. as per @zzmp

**Comment by Nex-Pro** [comment]
### 100% PASS 

Perfect.  Sound was clean. No hesitation on looping. Tested in desktop and HMD  static and moving. 
Tested very large wav file and smaller file.

No crackle hiss or pop.

Nice.

**Pull request merged**